modules/pipeline/providers/leaderworker/leader_listen_worker_task.go

```diff
@@ -168,10 +169,66 @@ func (p *provider) listWorkerTasks(ctx context.Context, workerID worker.ID) ([]w
 	}
 	var tasks []worker.LogicTask
 	for _, kv := range resp.Kvs {
-		logicTaskID := p.getWorkerTaskLogicIDFromIncomingKey(workerID, string(kv.Key))
+		logicTaskID := p.getWorkerLogicTaskIDFromIncomingKey(workerID, string(kv.Key))
 		logicTaskData := kv.Value
 		task := worker.NewLogicTask(logicTaskID, logicTaskData)
 		tasks = append(tasks, task)
 	}
 	return tasks, nil
 }
+
+func (p *provider) leaderListenTaskCanceling(ctx context.Context) {
+	prefix := p.makeEtcdLeaderLogicTaskCancelListenPrefix()
+	p.ListenPrefix(ctx, prefix,
+		func(ctx context.Context, event *clientv3.Event) {
+			// concurrent cancel
+			go func() {
+				key := string(event.Kv.Key)
+				logicTaskID := p.getLogicTaskIDFromLeaderCancelKey(key)
+				// check logic task
+				isHandling, workerID := p.IsTaskBeingProcessed(ctx, logicTaskID)
+				if !isHandling {
+					p.Log.Warnf("skip cancel logic task(not being processed), logicTaskID: %s", logicTaskID)
+					return
+				}
+				// do cancel
+				distributedCancelKey := p.makeEtcdWorkerLogicTaskCancelKey(workerID, logicTaskID)
+				for {
+					_, err := p.EtcdClient.Put(ctx, distributedCancelKey, "")
+					if err != nil {
+						p.Log.Errorf("failed to distributed cancel logic task(auto retry), workerID: %s, logicTaskID: %s, err: %v", workerID, logicTaskID, err)
+						time.Sleep(p.Cfg.Leader.RetryInterval)
+						continue
+					}
+					break
+				}
+			}()
+		},
+		nil,
+	)
+}
+
+func (p *provider) loadCancelingTasks(ctx context.Context) {
+	p.Log.Infof("begin load canceling logic tasks")
+	defer p.Log.Infof("end load canceling logic tasks")
+	prefix := p.makeEtcdLeaderLogicTaskCancelListenPrefix()
+	rutil.ContinueWorking(ctx, p.Log, func(ctx context.Context) rutil.WaitDuration {
+		getResp, err := p.EtcdClient.Get(ctx, prefix, clientv3.WithPrefix())
+		if err != nil {
+			p.Log.Errorf("failed to load canceling logic tasks from etcd(auto retry), err: %v", err)
+			return rutil.ContinueWorkingWithDefaultInterval
+		}
+		// put into etcd again to trigger cancel
+		for _, kv := range getResp.Kvs {
+			logicTaskID := p.getLogicTaskIDFromLeaderCancelKey(string(kv.Key))
+			_, err := p.EtcdClient.Put(ctx, string(kv.Key), string(kv.Value))
+			if err != nil {
+				p.Log.Errorf("failed to put into etcd when load canceling logic task, logicTaskID: %s, err: %v", logicTaskID, err)
+				return rutil.ContinueWorkingWithDefaultInterval
+			}
+			p.Log.Infof("load canceling logic task success, logicTaskID: %s", logicTaskID)
+			continue
+		}
```

Comment:
continue is unnecessary
is there a situation that worker delete cancel key, but leader put it again?

Reply:
worker delete cancel key only when pipeline id reconciler done.
It's idempotent even if leader put cancel key again after pipeline in end status.